### PR TITLE
Add support for rich push notifications

### PR DESCRIPTION
**Pull request by jbwyme**

### CODE REVIEW

File: Mixpanel/MixpanelNotificationServiceExtension.swift [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        self.maybeAttachButtons(bestAttemptContent: bestAttemptContent) {
+            self.maybeAttachMedia(bestAttemptContent: bestAttemptContent) {
+                contentHandler(bestAttemptContent)
+            }
+        }
+    }
+    
+    func maybeAttachButtons(bestAttemptContent: UNMutableNotificationContent, completionHandler: @escaping () -> Void) {
+        guard let buttons = bestAttemptContent.userInfo["mp_buttons"] as? [[AnyHashable: Any]] else {
+            NSLog("maybeAttachButtons: No action buttons found in the push notification payload.")
```
Comment (jbwyme) [inline Mixpanel/MixpanelNotificationServiceExtension.swift]: I realize that `NSLog` might not be the correct way to log this debugging info but trying to use `print` or the `MPLog` stuff resulted in nothing showing up. `NSLog` was the only way I was able to see output in the console.

File: Mixpanel/MixpanelNotificationServiceExtension.swift [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+    var contentHandler: ((UNNotificationContent) -> Void)?
+    var bestAttemptContent: UNMutableNotificationContent?
+
+    open override func didReceive(_ request: UNNotificationRequest, withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void) {
+
+        NSLog("%@ MPNotificationServiceExtension didReceiveNotificationRequest", self);
+
+        guard MixpanelPushNotifications.isMixpanelPushNotification(request.content) else {
+            NSLog("%@ Not a Mixpanel push notification, returning original content", self);
```
Comment (zihejia) [inline Mixpanel/MixpanelNotificationServiceExtension.swift]: Can we use Logger.info in here?

Comment (jbwyme) [reply Mixpanel/MixpanelNotificationServiceExtension.swift]: No, my understanding is that this is a completely new environment which means it's a different Mixpanel instance. Thus, we'd have to require customers to explicitly create a Mixpanel instance here and enable logging to get these. My thoughts were that this is a pretty rare thing so always `NSLog`ing won't pollute the logs much.

File: Mixpanel/PushNotifications.swift [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                guard let buttonOnTap = button["ontap"] as? [AnyHashable: Any] else {
+                    Logger.debug(message: "Expected 'ontap' property in button dict")
+                    completionHandler();
+                    return
+                }
+
+                ontap = buttonOnTap
+
+                trackingProps["tap_target"] = "button"
+                trackingProps["button_id"] = button["id"] as! String
```
Comment (zihejia) [inline Mixpanel/PushNotifications.swift]: prefer safely unwrap the optional in here

File: Mixpanel/MixpanelNotificationServiceExtension.swift [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        let buttons = content.userInfo["mp_buttons"] as? [[AnyHashable: Any]] ?? []
+        if (buttons.count == 0) {
+            NSLog("getCategoryIdentifier: No action buttons found in the push notification payload.")
+        }
+        
+        // Build a list of actions from the buttons data
+        var actions: [UNNotificationAction] = []
+        for (idx, button) in buttons.enumerated() {
+            let identifier = String(format: "MP_ACTION_%lu", idx)
+            let title = button["lbl"] as! String
```
Comment (zihejia) [inline Mixpanel/MixpanelNotificationServiceExtension.swift]: prefer safely unwrap the optional in here

File: Mixpanel/MixpanelNotificationServiceExtension.swift [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        self.getCategoryIdentifier(content: request.content) { categoryIdentifier in
+
+            if let categoryIdentifier = categoryIdentifier {
+                NSLog("Using categoryIdentifer: \(categoryIdentifier)")
+                bestAttemptContent.categoryIdentifier = categoryIdentifier
+            }
+
+            // Download rich media and create an attachment
+            self.buildAttachments(content: request.content) { attachments in
+                if attachments != nil {
```
Comment (zihejia) [inline Mixpanel/MixpanelNotificationServiceExtension.swift]:
```                if let attachments = attachments {
                    NSLog("Adding \(attachments.count) attachment(s)")
                    bestAttemptContent.attachments = attachments
                }
```

File: Mixpanel/PushNotifications.swift [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            return;
+        }
+
+        var ontap: [AnyHashable: Any]? = nil
+
+        if response.actionIdentifier == UNNotificationDefaultActionIdentifier {
+            // The action that indicates the user opened the app from the notification interface.
+            trackingProps["tap_target"] = "notification";
+
+            if (userInfo["mp_ontap"] != nil) {
```
Comment (zihejia) [inline Mixpanel/PushNotifications.swift]: nit: if let ..., 
so you don't have to force wrap.

File: MixpanelDemo/NotificationServiceExtension/NotificationService.swift [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+//  NotificationService.swift
+//  NotificationServiceExtension
+//
+//  Created by Josh Wymer on 1/18/20.
+//  Copyright © 2020 Mixpanel. All rights reserved.
+//
+
+import Mixpanel
+
+class NotificationService: MixpanelNotificationServiceExtension  {}
```
Comment (zihejia) [inline MixpanelDemo/NotificationServiceExtension/NotificationService.swift]: is this class being used?

File: Mixpanel/MixpanelNotificationServiceExtension.swift [hunk 8]
```diff
@@ -0,0 +1,168 @@
+import UserNotifications
+
+@available(iOS 11.0, *)
+open class MixpanelNotificationServiceExtension: UNNotificationServiceExtension {
+
+    var contentHandler: ((UNNotificationContent) -> Void)?
+    var bestAttemptContent: UNMutableNotificationContent?
+
+    open override func didReceive(_ request: UNNotificationRequest, withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void) {
+
```
Comment (zihejia) [inline Mixpanel/MixpanelNotificationServiceExtension.swift]: nit, remove empty line

File: Mixpanel/MixpanelNotificationServiceExtension.swift [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        }
+
+        // Store a reference to the mutable content and the contentHandler on the class so we
+        // can use them in serviceExtensionTimeWillExpire if needed
+        self.contentHandler = contentHandler
+        self.bestAttemptContent = bestAttemptContent
+
+        // Setup the category first since it's faster and less likely to cause time to expire
+        self.getCategoryIdentifier(content: request.content) { categoryIdentifier in
+
```
Comment (zihejia) [inline Mixpanel/MixpanelNotificationServiceExtension.swift]: nit, remove empty line

File: Mixpanel/MixpanelNotificationServiceExtension.swift [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        guard let bestAttemptContent = self.bestAttemptContent else {
+            return;
+        }
+
+        contentHandler(bestAttemptContent);
+    }
+    
+    func getCategoryIdentifier(content: UNNotificationContent, completionHandler: @escaping (String?) -> Void) {
+
```
Comment (zihejia) [inline Mixpanel/MixpanelNotificationServiceExtension.swift]: nit, remove empty line

File: Mixpanel/PushNotifications.swift (hunk 6 above)
Comment (jbwyme) [reply Mixpanel/PushNotifications.swift]: we don't actually force unwrap in this instance

**Pull request merged**